### PR TITLE
 	MultiPort support (-p 8888 -p 8388 -p soon upto 20 port) and fix buffer overflow in -s

### PR DESCRIPTION
**Pull request by KrazyIvan-**
Fix buffer overflow on start up: server_num  and remote_num  to not over MAX_REMOTE_NUM

### CODE REVIEW

**Comment by madeye** [comment]
Please squash your commits first.

http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

**Comment by madeye** [comment]
Also, please follow the code style defined here: https://github.com/shadowsocks/shadowsocks-libev/blob/master/.uncrustify.cfg

File: src/jconf.c [hunk 1]
```diff
@@ -153,7 +150,16 @@ jconf_t *read_jconf(const char * file)
                     conf.remote_num = 1;
                 }
             } else if (strcmp(name, "server_port") == 0) {
-                conf.remote_port = to_string(value);
+				 if (value->type == json_array) {
```
Comment (madeye) [inline src/jconf.c]: Fix the indent.

File: src/jconf.c [hunk 2]
```diff
@@ -153,7 +150,16 @@ jconf_t *read_jconf(const char * file)
...
+                    for (j = 0; j < value->u.array.length && j < MAX_PORT_NUM; j++) {                        
+                        json_value *v = value->u.array.values[j];
+						conf.remote_port[j] = to_string(v);
+                        conf.port_num = j + 1;
+                    }
+                } else if (value->type == json_string) {
+					conf.remote_port[0] = to_string(value);
```
Comment (madeye) [inline src/jconf.c]: Ditto.

File: src/local.c [hunk 3]
```diff
@@ -911,8 +911,10 @@ int main(int argc, char **argv)
             }
             break;
         case 's':
-            remote_addr[remote_num].host = optarg;
-            remote_addr[remote_num++].port = NULL;
+			if (remote_num < MAX_REMOTE_NUM) {
```
Comment (madeye) [inline src/local.c]: Ditto.

File: src/redir.c [hunk 4]
```diff
@@ -613,8 +613,10 @@ int main(int argc, char **argv)
     while ((c = getopt(argc, argv, "f:s:p:l:k:t:m:c:b:a:")) != -1) {
         switch (c) {
         case 's':
-            remote_addr[remote_num].host = optarg;
-            remote_addr[remote_num++].port = NULL;
+			if (remote_num < MAX_REMOTE_NUM) {
```
Comment (madeye) [inline src/redir.c]: Ditto.

File: src/server.c [hunk 5]
```diff
@@ -1043,10 +1044,14 @@ int main(int argc, char **argv)
             }
             break;
         case 's':
-            server_host[server_num++] = optarg;
+			if (server_num < MAX_REMOTE_NUM) {
```
Comment (madeye) [inline src/server.c]: Ditto.

File: src/server.c [hunk 6]
```diff
@@ -1199,44 +1207,48 @@ int main(int argc, char **argv)
...
-        listen_ctx->iface = iface;
-        listen_ctx->loop = loop;
-
-        ev_io_init(&listen_ctx->io, accept_cb, listenfd, EV_READ);
-        ev_io_start(loop, &listen_ctx->io);
-    }
+	int port_i;
+	int server_i;
+	for(port_i=0;port_i<port_num;port_i++){
+		for(server_i=0;server_i<server_num;server_i++) {
```
Comment (madeye) [inline src/server.c]: Ditto.

File: src/server.c [hunk 7]
```diff
@@ -1199,44 +1207,48 @@ int main(int argc, char **argv)
...
+			ev_io_start(loop, &listen_ctx->io);
+		}
+	}
 
     // Setup UDP
     if (udprelay) {
         LOGI("udprelay enabled");
-        init_udprelay(server_host[0], server_port, m, atoi(timeout),
-                      iface);
+		for(port_i=0;port_i<port_num;port_i++) {
```
Comment (madeye) [inline src/server.c]: You cannot do that. UDP relay only works on one port.

Comment (KrazyIvan-) [inline src/server.c]: then just:
    if (udprelay) {
        LOGI("udprelay enabled");		
		init_udprelay(server_host[0], server_port[0], m, atoi(timeout),
                     iface);
		
    }

I need to fix code there also

Comment (KrazyIvan-) [inline src/server.c]: Squashing is not support via web

Comment (madeye) [inline src/server.c]: Follow the steps in this link: http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

Comment (KrazyIvan-) [inline src/server.c]: tcp6       0      0 :::1000                 :::*                    LISTEN      2924/src/ss-server
tcp6       0      0 :::1001                 :::*                    LISTEN      2924/src/ss-server
tcp6       0      0 :::1002                 :::*                    LISTEN      2924/src/ss-server
udp6       0      0 ::1:1000                :::*                                2924/src/ss-server
udp6       0      0 ::1:1001                :::*                                2924/src/ss-server
udp6       0      0 ::1:1002                :::*                                2924/src/ss-server

its working with multi  UDP

Comment (madeye) [inline src/server.c]: Multiple UDP contexts cannot be freed correctly with free_udprelay().

Comment (madeye) [inline src/server.c]: You need to also modify udprelay.c to support this feature.

**Pull request closed without merging**